video/out/vo_sixel.c

```diff
@@ -0,0 +1,429 @@
+#include "osdep/terminal.h"
+#include "sub/osd.h"
+#include "vo.h"
+#include "video/sws_utils.h"
+#include "video/mp_image.h"
+
+#define IMGFMT IMGFMT_RGB24
+
+#define ESC_HIDE_CURSOR "\033[?25l"
+#define ESC_RESTORE_CURSOR "\033[?25h"
+#define ESC_CLEAR_SCREEN "\033[2J"
+#define ESC_GOTOXY "\033[%d;%df"
+#define ESC_TERMINATE_DCS "\033\\"
+#define ESC_SAVE_XY_CURSOR_LOCATION "\033[%d;%dH\0337"
+#define ESC_RESTORE_SAVED_CURSOR "\0338"
+#define ESC_USE_GLOBAL_COLOR_REG "\033[?1070l"
+
+
+struct vo_sixel_opts {
+    int diffuse;
+    int width;
+    int height;
+    int reqcolors;
+    int fixedpal;
+    int threshold;
+    int top;
+    int left;
+};
+
+#define OPT_BASE_STRUCT struct vo_sixel_opts
+static const struct m_sub_options vo_sixel_conf = {
+    .opts = (const m_option_t[]) {
+        {"vo-sixel-diffusion", OPT_CHOICE(diffuse,
+            {"auto", DIFFUSE_AUTO},
+            {"none", DIFFUSE_NONE},
+            {"atkinson", DIFFUSE_ATKINSON},
+            {"fs", DIFFUSE_FS},
+            {"jajuni", DIFFUSE_JAJUNI},
+            {"stucki", DIFFUSE_STUCKI},
+            {"burkes", DIFFUSE_BURKES},
+            {"arithmetic", DIFFUSE_A_DITHER},
+            {"xor", DIFFUSE_X_DITHER})},
+        {"vo-sixel-width", OPT_INT(width)},
+        {"vo-sixel-height", OPT_INT(height)},
+        {"vo-sixel-reqcolors", OPT_INT(reqcolors)},
+        {"vo-sixel-fixedpalette", OPT_INT(fixedpal)},
+        {"vo-sixel-color-threshold", OPT_INT(threshold)},
+        {"vo-sixel-offset-top", OPT_INT(top)},
+        {"vo-sixel-offset-left", OPT_INT(left)},
+        {0}
+    },
+    .defaults = &(const struct vo_sixel_opts) {
+        .diffuse = DIFFUSE_ATKINSON,
+        .width = 320,
+        .height = 240,
+        .reqcolors = 256,
+        .fixedpal = 0,
+        .threshold = 0,
+        .top = 1,
+        .left = 1,
+    },
+    .size = sizeof(struct vo_sixel_opts),
+};
+
+struct priv {
+    struct vo_sixel_opts *opts;
+
+    sixel_output_t *output;
+    sixel_dither_t *dither;
+    sixel_dither_t *testdither;
+    uint8_t        *buffer;
+
+    int image_height;
+    int image_width;
+    int image_format;
+
+    struct mp_image *frame;
+    struct mp_sws_context *sws;
+};
+
+static const unsigned int depth = 3;
+
+static void save_image_offset_location(struct vo* vo)
+{
+    struct priv* priv = vo->priv;
+    int top = priv->opts->top;
+    int left = priv->opts->left;
+    int terminal_width = 0;
+    int terminal_height = 0;
+
+    terminal_get_size(&terminal_width, &terminal_height);
+
+    // Make sure that the user specified top offset
+    // lies in the range 1 to TERMINAL_HEIGHT
+    // Otherwise default to the topmost row
+    if (top <= 0 || top > terminal_height)
+        top = 1;
+
+    // Make sure that the user specified left offset
+    // lies in the range 1 to TERMINAL_WIDTH
+    // Otherwise default to the leftmost column
+    if (left <= 0 || top > terminal_width)
+        left = 1;
+
+    /*We want to save the offset row and column
```

Comment:
Why do we want to save it? You know exactly where the image should start - either given by the offset options or because you calculate it yourself, at which case you don't need to save it by the terminal - you know the values and can just move the cursor yourself to that location, no?

Reply:
Yes I can replace this function with the `GOTOXY` sequence. But I think it requires printing more characters instead of just two bytes in `\0338` which might be insignificant per frame.  The saving of coordinates happens just once per resize function callback so to me from mpv perspective, it should be more efficient to just restore.

Reply:
It indeed completely insignificant compared to the image data.

I think I prefer explicit move of the cursor, because otherwise you can't really tell what the terminal would do, especially if scroll happened between the save and restore. Let's make it explicit.